src/axolotl/utils/callbacks.py

```diff
@@ -317,3 +320,250 @@ def on_evaluate(
+                    
+                    # # Extract prompt and completion tokens from input_ids based on labels
+                    # prompt_token_ids = batch.data['input_ids'][batch.data['labels'] == IGNORE_INDEX]
+                    # completion_token_ids = batch.data['input_ids'][batch.data['labels'] != IGNORE_INDEX]
+
+                    # # prompt_texts = tokenizer.batch_decode(batch.data['input_ids'])
+                    # prompt_texts = tokenizer.batch_decode(prompt_token_ids)
+                    # completion_texts = tokenizer.batch_decode(completion_token_ids)
+
+                    (loss, logits, labels) = trainer.prediction_step(
+                        trainer.model,
+                        batch,
+                        prediction_loss_only=False,
+                    )
+
+                    # prompt_completion_pairs = zip(prompt_texts, logits)
+
+                    # print("logits", logits)
+                    # print("labels", labels)
+
+                    # pred_tokens = []
+                    # for i, logit in enumerate(logits):
+                    for i, (logit, labels_i) in enumerate(zip(logits, labels)):
+                        # for i, (prompt_text, logit) in enumerate(prompt_completion_pairs):
+                        # print(dir(logit))
+                        # print(logit)
+                        # print(logit.shape)
+                        # # Convert the logits to probabilities using softmax
+                        # probabilities = torch.softmax(logit, dim=-1)
+
+                        # # Get the predicted token id (the one with the highest probability)
+                        # predicted_token_id = torch.argmax(probabilities).item()
+
+                        # # Decode the predicted token id to get the plaintext
+                        # predicted_token = tokenizer.decode([predicted_token_id])
+
+                        # # Append the predicted token to the preds list
+                        # pred_tokens.append(predicted_token)
+
+                        # # Convert the logits to probabilities using softmax
+                        # probabilities = torch.softmax(logit, dim=-1)
+
+                        # # Get the predicted token ids (the ones with the highest probability)
+                        # predicted_token_ids = torch.argmax(probabilities, dim=-1)
+
+                        # # Decode the predicted token ids to get the plaintext
+                        # predicted_tokens = tokenizer.batch_decode(predicted_token_ids)
+
+                        # 
+                        # label_non_zero_indices = (batch["labels"][i] != IGNORE_INDEX).nonzero().transpose(0, 1)[0] # FIXME: clean up?
+
+                        prompt_token_indices = (batch["labels"][i] == IGNORE_INDEX).nonzero().transpose(0, 1)[0] # FIXME: clean up?
+                        completion_token_indices = (batch["labels"][i] != IGNORE_INDEX).nonzero().transpose(0, 1)[0] # FIXME: clean up?
```

Comment:
**Help Wanted**
I don't think this is the correct way to extract the original tokens for 1️⃣  prompt and 2️⃣ expected completion for each example in the `eval_dataset`.

Reply:
Note I also had to set `sample_packing: false` for this to even appear to work properly.

Reply:
yeah, you could also use the positiion_ids to split up the packed sequences, but using the ignore index is how I'd do it too

Reply:
With `sample_packing: true` then `batch['position_ids']` appeared. I was able to use it to extract out the individual prompts within each of the packed batch entries ✅  Thanks for recommendation!

I continued using the `labels` for `sample_packing: false` as you'll see.